### PR TITLE
Android Issue #225 Hilt Set up

### PR DESCRIPTION
**Pull request by jyoshna-joshi**
Closes issue #225 

**Changed files:**
TaskDetailScreen/TaskDetailViewModel- implemented hilt
TaskListScreen/TaskListViewModel  - factory removed, implemented  hilt

**New files added:**
Datamodules folder - for providing dependencies 
ITaskRepository - Task Repository interface for defining methods
TaskApplication - for hilt

@baghaii @gaurigadkari @renweview My PR is ready for review !

### CODE REVIEW

**Comment by jyoshna-joshi** [comment]
Hi @baghaii, I'd like to hold off until #236 gets merged. This way, she won't have to deal with conflicts arising from this pull request. I will re-work on it.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/AppDatabase.kt [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+import androidx.room.Database
 import androidx.room.RoomDatabase
+import com.example.tasktracker.data.model.Task
 
 /**
  * Created by Gauri Gadkari on 1/23/24.
  * The Room database for this app
  */
-
+@Database(entities = [Task::class], version = 1, exportSchema = false)
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/AppDatabase.kt:11]: I think this is created in @SttApollo 's PR so it is not needed here, but you may want to remove WHERE the database gets created. I think you have it created in the dependency injection, and she has it created in this file. Your approach is cleaner when hilt is available.

https://github.com/WomenWhoCode/WWCodeMobile/pull/236/files#diff-61b73221ef21cb31994fc5760de2ff233f965b43d4b93f7f4f9a80104b027466

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskDao.kt [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 
 /**
  * Created by Gauri Gadkari on 1/23/24.
  */
 @Dao
 interface TaskDao {
-    abstract fun taskDao(): TaskDao
+    //todo replace with required method
+    @Update
+    suspend fun updateTask(task: Task)
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskDao.kt:14]: This was also created in @SttApollo's PR, and this one is not needed.

xhttps://github.com/WomenWhoCode/WWCodeMobile/pull/236/files#diff-7fadad1694f917f3b9f439b227faeea1b0f5569dbd7e3ec0f263ed06b7260de9

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/di/DataModules.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            context.applicationContext,
+            AppDatabase::class.java,
+            "tasklist.db"
+        ).build()
+    }
+}
+
+@Module
+@InstallIn(SingletonComponent::class)
+abstract class RepositoryModule{
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/di/DataModules.kt:41]: I would probably put this module in a separate file.

**Comment by baghaii** [comment]
This is much easier to review because it is smaller, and it is going to be a great addition!

Both @gaurigadkari and I are a little concerned about the JDK upgrade because it might harm other people's ability to contribute to the project. Do you remember why it was needed?

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/AppDatabase.kt (hunk 1 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/AppDatabase.kt:11]: I will remove it once her PR gets merged.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskDao.kt (hunk 2 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskDao.kt:14]: I will remove this one.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/di/DataModules.kt (hunk 3 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/di/DataModules.kt:41]: Noted. Will do it.

**Comment by jyoshna-joshi** [comment]
Screenshot 2024-03-10 104754

JDK update was done because of this issue.  However, if we don't want to update it. We can add following in gradle :
image

Following is the link I followed:
https://stackoverflow.com/questions/76030538/android-agp-8-gradle-8-kotlin-1-8-causes-error-in-kapt

**Comment by jyoshna-joshi** [comment]
Hi @baghaii, Please have a look into my updated changes.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt [hunk 4]
```diff
@@ -57,7 +58,7 @@ import java.util.TimeZone
  */
     @OptIn(ExperimentalMaterial3Api::class)
     @Composable
-    fun TaskDetailScreen(onNavigateToList: () -> Unit) {
+    fun TaskDetailScreen(onNavigateToList: () -> Unit, viewModel: TaskDetailViewModel = hiltViewModel()) {
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt:61]: I wasn't sure how or where to set up these view models so I spoke to someone earlier this week, and she suggested following the example in the documentation.

https://developer.android.com/jetpack/compose/libraries#hilt-navigation

These set up the view model, in the navigation and then pass them to the screens. 
Both task detail and task list should be set up like this.

**Comment by baghaii** [comment]
@jyoshna-joshi I went ahead and merged the Room PR so you can fix the merge conflicts here.

@sierraobryan left a last set of comments on the Room PR. Do you think you can incorporate some of the changes that she requested in this PR. 

I know that I requested that you move the creation of the database. Can you rename the database to something more relevant to this project when you do that? 

Could we update to use suspend functions instead of LiveData?

https://github.com/WomenWhoCode/WWCodeMobile/pull/236/files

**Comment by jyoshna-joshi** [comment]
@baghaii  I will work on this one.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainActivity.kt [hunk 5]
```diff
@@ -3,19 +3,14 @@ package com.example.tasktracker
...
-import androidx.compose.ui.res.painterResource
-import androidx.compose.ui.tooling.preview.Preview
-import com.example.tasktracker.ui.TaskList.TaskListScreen
 import com.example.tasktracker.ui.theme.TaskTrackerTheme
+import dagger.hilt.android.AndroidEntryPoint
 
+/**
+ * Integrated Hilt by Jyoshna on 8/3/24
+ */
+
```
Comment (gaurigadkari) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainActivity.kt:12]: nit: can we remove this comment from here, at a comment above the class name should tell us more about the class

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/di/RepositoryModules.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package com.example.tasktracker.di
+
+import com.example.tasktracker.data.ITaskRepository
+import com.example.tasktracker.data.TaskRepository
+import dagger.Binds
+import dagger.Module
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+/**
```
Comment (gaurigadkari) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/di/RepositoryModules.kt:10]: nit: can you reformat all the files, there is an inconsistency with linebreaks

File: coding-projects/android/TaskTracker/build.gradle.kts [hunk 7]
```diff
@@ -2,4 +2,5 @@
 plugins {
     id("com.android.application") version "8.1.1" apply false
     id("org.jetbrains.kotlin.android") version "1.8.10" apply false
+    id("com.google.dagger.hilt.android") version "2.44" apply false
```
Comment (gaurigadkari) [inline coding-projects/android/TaskTracker/build.gradle.kts:5]: do we not need to add:
dependencies {
  implementation("com.google.dagger:hilt-android:2.44")
  kapt("com.google.dagger:hilt-android-compiler:2.44")
}

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt (hunk 4 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt]: I implemented this, could you please check @baghaii ?

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainActivity.kt (hunk 5 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainActivity.kt]: It's removed.

File: coding-projects/android/TaskTracker/build.gradle.kts (hunk 7 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/build.gradle.kts:5]: Currently, it is not working without them. I will have deeper look.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainScreen.kt [hunk 8]
```diff
@@ -1,30 +1,33 @@
...
+            TaskListScreen(onNavigateToSettings = { navController.navigate(NavScreens.TaskSettings.route) },
                 onNavigateToDetail = { navController.navigate(NavScreens.TaskDetail.route) })
         }
         composable(NavScreens.TaskSettings.route) {
             TaskSettingsScreen()
         }
         composable(NavScreens.TaskDetail.route) {
-            TaskDetailScreen(onNavigateToList = { navController.navigate(NavScreens.TaskList.route) },
-                )
+            val viewModel = hiltViewModel<TaskDetailViewModel>()
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainScreen.kt:26]: Yup. This is what Sierra showed me in the examples. Do we have a view model for the list screen that we can inject for that one too?

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskRepository.kt [hunk 9]
```diff
@@ -1,50 +1,68 @@
...
     //Static list for testing
     companion object {
         // Static list for testing
         val tasks = listOf(
-            Task(id = 1, activityName = "Morning Walk", date = "2024-02-26", startTimeInMillis = 7 * 60 * 60 * 1000, endTimeInMillis = 8 * 60 * 60 * 1000, duration = "1h 0m"),
-            Task(id = 2, activityName = "Read Book", date = "2024-02-26", startTimeInMillis = 9 * 60 * 60 * 1000, endTimeInMillis = 10 * 60 * 60 * 1000, duration = "1h 0m"),
-            Task(id = 3, activityName = "Online Course", date = "2024-02-26", startTimeInMillis = 11 * 60 * 60 * 1000, endTimeInMillis = 13 * 60 * 60 * 1000, duration = "2h 0m"),
-            Task(id = 4, activityName = "Lunch Break", date = "2024-02-26", startTimeInMillis = 13 * 60 * 60 * 1000, endTimeInMillis = 14 * 60 * 60 * 1000, duration = "1h 0m"),
-            Task(id = 5, activityName = "Project Work", date = "2024-02-26", startTimeInMillis = 15 * 60 * 60 * 1000, endTimeInMillis = 18 * 60 * 60 * 1000, duration = "3h 0m")
+            Task(
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskRepository.kt:30]: Were these being used at all? In my version of the app, they looked unused. Could we delete this companion object?

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailViewModel.kt [hunk 10]
```diff
@@ -1,22 +1,20 @@
...
-class TaskDetailViewModel(application: Application) : AndroidViewModel(application) {
-
-    private val repository: TaskRepository = TaskRepository(application)
-
-    fun insertTask(newTask: Task) {
-        repository.insertTask(newTask)
+@HiltViewModel
+class TaskDetailViewModel @Inject constructor(private var respository: TaskRepository) :
+    ViewModel() {
+    val allTasks: LiveData<List<Task>>? = respository.allTasks
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailViewModel.kt:16]: I don't think we need the list of all of them here for the detail screen. We need these probably in the TaskListViewModel.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt (hunk 4 above)
Comment (baghaii) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt]: This is looking good. @gaurigadkari and @jyoshna-joshi , are you able to run the version of the code on this branch without problems?

Of the comments I made, the ones that feel most needed are:
- Have all the viewmodels right before the composables in the navigation structure of main screen.
- Try to strip out some of the lists of fake data we have on the various screens if possible.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskList/TaskListViewModel.kt [hunk 11]
```diff
@@ -4,14 +4,18 @@ import androidx.lifecycle.ViewModel
...
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.flowOf
+import javax.inject.Inject
 
 /**
  * Created by Gauri Gadkari on 1/23/24.
  */
-class TaskListViewModel: ViewModel() {
+@HiltViewModel
+class TaskListViewModel @Inject constructor(private var respository: TaskRepository) : ViewModel() {
```
[the remark this answers is not in the document]
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskList/TaskListViewModel.kt:18]: For populating real list, I think we require  #73.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailViewModel.kt (hunk 10 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailViewModel.kt]: It has been removed. Insert, update and delete are available here.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskRepository.kt (hunk 9 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskRepository.kt]: I have removed this code.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskRepository.kt [hunk 12]
```diff
@@ -1,50 +1,68 @@
...
 
 /**
  * Created by Gauri Gadkari on 1/23/24.
  */
-class TaskRepository(application: Application) {
-    private var db: AppDatabase? = null
-    private var taskDAO = AppDatabase.getDB(application)?.taskDAO()
-    var allTasks: LiveData<List<Task>>? = taskDAO?.getAllTasks()
+class TaskRepository @Inject constructor(private val taskDAO: TaskDAO) : ITaskRepository {
+    var allTasks: LiveData<List<Task>>? = taskDAO.getAllTasks()
```
[the remark this answers is not in the document]
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/data/TaskRepository.kt]: Flow is integrated now.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainScreen.kt (hunk 8 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/MainScreen.kt:31]: It's done in list screen as well.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt (hunk 4 above)
Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskDetail/TaskDetailScreen.kt]: Current version of code is running good without problems.

**Comment by jyoshna-joshi** [comment]
@baghaii Please have a look into new changes.

File: coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskList/TaskListViewModel.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (baghaii) [reply coding-projects/android/TaskTracker/app/src/main/java/com/example/tasktracker/ui/TaskList/TaskListViewModel.kt:18]: Fair point.

File: coding-projects/android/TaskTracker/app/src/androidTest/java/com/example/tasktracker/AppDatabaseTest.kt [hunk 13]
```diff
@@ -56,37 +59,31 @@ class AppDatabaseTest {
...
-            data = o
-            latch.countDown()
-        }
-
-        InstrumentationRegistry.getInstrumentation().runOnMainSync {
-            liveData.observeForever(observer)
+        val mainScope = MainScope()
+        mainScope.launch {
+            flowData
+                .flowOn(Dispatchers.Main)
```
Comment (baghaii) [inline coding-projects/android/TaskTracker/app/src/androidTest/java/com/example/tasktracker/AppDatabaseTest.kt:73]: I believe that we do database things on Dispatchers.IO.

Usually, we test in a TestScope, but I was having a tough time trying to make that work. I am glad you got the test working.

Comment (jyoshna-joshi) [reply coding-projects/android/TaskTracker/app/src/androidTest/java/com/example/tasktracker/AppDatabaseTest.kt]: It's done.

**Comment by jyoshna-joshi** [comment]
@baghaii Conflicts resolved.

**Review by baghaii — APPROVED** [review]

**Pull request merged**